ext/ddtrace_profiling_native_extension/collectors_cpu_and_wall_time_worker.c

```diff
@@ -1053,3 +1063,20 @@ static VALUE rescued_sample_allocation(VALUE tracepoint_data) {
   // Return a dummy VALUE because we're called from rb_rescue2 which requires it
   return Qnil;
 }
+
+static void stop_if_allocation_sampling_failure(struct cpu_and_wall_time_worker_state *state) {
+  const char* error = discrete_dynamic_sampler_error(&state->allocation_sampler);
+  if (error != NULL) {
+    VALUE error_rbstr = rb_str_new_cstr("allocation sampler - ");
+    rb_str_cat_cstr(error_rbstr, discrete_dynamic_sampler_error(&state->allocation_sampler));
+    stop_state(state, rb_exc_new_str(rb_eRuntimeError, error_rbstr));
+  }
+}
```

Comment:
Having to remember to check the error is a bit awkward. One technique we use for libdatadog, and for instance in https://github.com/DataDog/dd-trace-rb/blob/master/ext/ddtrace_profiling_native_extension/clock_id.h is to return a struct from the APIs that immediately flags if there was an error or not.

Wouldn't say it's a blocker for this PR, but may be cool to experiment with maybe?